pkg/resourcemanager/mysql/mysqluser/mysqluser_reconcile.go

```diff
@@ -0,0 +1,319 @@
+)
+
+// Ensure that user exists
+func (s *MySqlUserManager) Ensure(ctx context.Context, obj runtime.Object, opts ...resourcemanager.ConfigOption) (bool, error) {
+	instance, err := s.convert(obj)
+	if err != nil {
+		return false, err
+	}
+
+	requestedUsername := instance.Spec.Username
+	if len(requestedUsername) == 0 {
+		requestedUsername = instance.Name
+	}
+
+	options := &resourcemanager.Options{}
+	for _, opt := range opts {
+		opt(options)
+	}
+
+	adminSecretClient := s.SecretClient
+
+	adminsecretName := instance.Spec.AdminSecret
+	if len(instance.Spec.AdminSecret) == 0 {
+		adminsecretName = instance.Spec.Server
+	}
+
+	key := types.NamespacedName{Name: adminsecretName, Namespace: instance.Namespace}
+
+	var mysqlUserSecretClient secrets.SecretClient
+	if options.SecretClient != nil {
+		mysqlUserSecretClient = options.SecretClient
+	} else {
+		mysqlUserSecretClient = s.SecretClient
+	}
+
+	// if the admin secret keyvault is not specified, fall back to global secretclient
+	if len(instance.Spec.AdminSecretKeyVault) != 0 {
+		adminSecretClient = keyvaultSecrets.New(instance.Spec.AdminSecretKeyVault)
+		if len(instance.Spec.AdminSecret) != 0 {
+			key = types.NamespacedName{Name: instance.Spec.AdminSecret}
+		}
+	}
+
+	// get admin creds for server
+	adminSecret, err := adminSecretClient.Get(ctx, key)
+	if err != nil {
+		instance.Status.Provisioning = false
+		instance.Status.Message = fmt.Sprintf("admin secret : %s, not found in %s", key.String(), reflect.TypeOf(adminSecretClient).Elem().Name())
+		return false, nil
+	}
+
+	_, err = s.GetDB(ctx, instance.Spec.ResourceGroup, instance.Spec.Server, instance.Spec.DbName)
+	if err != nil {
+		instance.Status.Message = errhelp.StripErrorIDs(err)
+		instance.Status.Provisioning = false
+
+		requeuErrors := []string{
+			errhelp.ResourceNotFound,
+			errhelp.ParentNotFoundErrorCode,
+			errhelp.ResourceGroupNotFoundErrorCode,
+		}
+		azerr := errhelp.NewAzureErrorAzureError(err)
+		if helpers.ContainsString(requeuErrors, azerr.Type) {
+			return false, nil
+		}
+
+		// if the database is busy, requeue
+		errorString := err.Error()
+		if strings.Contains(errorString, "Please retry the connection later") {
+			return false, nil
+		}
+
+		// if this is an unmarshall error - igmore and continue, otherwise report error and requeue
+		if _, ok := err.(*json.UnmarshalTypeError); ok {
+			return false, err
+		}
+	}
+
+	adminUser := string(adminSecret["fullyQualifiedUsername"])
+	adminPassword := string(adminSecret[MSecretPasswordKey])
+	fullServerName := string(adminSecret["fullyQualifiedServerName"])
+
+	db, err := s.ConnectToSqlDb(ctx, MDriverName, fullServerName, instance.Spec.DbName, MSqlServerPort, adminUser, adminPassword)
+	if err != nil {
+		instance.Status.Message = errhelp.StripErrorIDs(err)
+		instance.Status.Provisioning = false
+
+		// catch firewall issue - keep cycling until it clears up
+		if strings.Contains(err.Error(), "is not allowed to connect to this MySQL server") {
```

Comment:
Should we log this somewhere or set some status in case this error condition persists? Or is there another way users can get visibility to this failure mode?

Reply:
Yes, we assign the error to instance.Status.Message just above this. And this helps the user see this when they do a kubectl describe of the instance

Reply:
Ha, silly me, was just being a little blind. Thanks.